sw/device/tests/dif/dif_plic_test.cc

```diff
@@ -0,0 +1,384 @@
+using mock_mmio::MockDevice;
+using testing::Test;
+
+class PlicTest : public Test, public MmioTest {
+ protected:
+  mmio_region_t base_addr_ = dev().region();
+  dif_plic_t dif_plic_ = {
+      /* base_addr = */ base_addr_,
+  };
+};
+
+class InitTest : public PlicTest {
+ protected:
+  void ExpectInitReset() {
+    // Interupt enable multireg.
+    EXPECT_WRITE32(RV_PLIC_IE00_REG_OFFSET, 0);
+    EXPECT_WRITE32(RV_PLIC_IE01_REG_OFFSET, 0);
+    EXPECT_WRITE32(RV_PLIC_IE02_REG_OFFSET, 0);
+
+    // Level/edge multireg.
+    EXPECT_WRITE32(RV_PLIC_LE0_REG_OFFSET, 0);
+    EXPECT_WRITE32(RV_PLIC_LE1_REG_OFFSET, 0);
+    EXPECT_WRITE32(RV_PLIC_LE2_REG_OFFSET, 0);
+
+    // Priority registers.
+    for (int i = 0; i < RV_PLIC_PARAM_NUMSRC; ++i) {
+      ptrdiff_t offset = RV_PLIC_PRIO0_REG_OFFSET + (sizeof(uint32_t) * i);
+      EXPECT_WRITE32(offset, 0);
+    }
+
+    // Target threshold registers.
+    EXPECT_WRITE32(RV_PLIC_THRESHOLD0_REG_OFFSET, 0);
+
+    // Software interrupt pending register.
+    EXPECT_WRITE32(RV_PLIC_MSIP0_REG_OFFSET, 0);
+  }
+};
+
+TEST_F(InitTest, NullArgs) {
+  dif_plic_result_t result = dif_plic_init(base_addr_, nullptr);
+  EXPECT_EQ(result, kDifPlicBadArg);
+}
+
+TEST_F(InitTest, Success) {
+  ExpectInitReset();
+
+  dif_plic_result_t result = dif_plic_init(base_addr_, &dif_plic_);
+  EXPECT_EQ(result, kDifPlicOk);
+}
+
+class IrqTests : public PlicTest {
+ protected:
+  struct Register {
+    ptrdiff_t offset;  // Register offset from the base.
+    uint8_t last_bit;  // Last bit index in the register.
+  };
+
+  // Set enable/disable multireg expectations, one bit per call.
+  void ExpectIrqSetTests(const std::vector<Register> &regs, bool enable) {
+    for (const auto &reg : regs) {
+      for (uint32_t i = 0; i <= reg.last_bit; ++i) {
+        EXPECT_MASK32(reg.offset, {{i, 0x1, enable}});
+      }
+    }
+  }
+
+  // Set multireg get status expectations, one bit per call.
+  void ExpectIrqGetTests(const std::vector<Register> &regs, bool enabled) {
+    for (const auto &reg : regs) {
+      for (int i = 0; i <= reg.last_bit; ++i) {
+        uint32_t value = 0x1 << i;
+
+        if (!enabled) {
+          value = ~value;
+        }
+
+        EXPECT_READ32(reg.offset, value);
+      }
+    }
+  }
+};
+
+class IrqEnableSetTest : public IrqTests {
+ protected:
+  // Make sure our tests are up to date.
+  IrqEnableSetTest() {
+    EXPECT_EQ(RV_PLIC_PARAM_NUMTARGET, 1);
+    EXPECT_EQ(registers_.size(), RV_PLIC_IE0_MULTIREG_COUNT);
+  }
+
+  std::vector<Register> registers_{
+      {
+          RV_PLIC_IE00_REG_OFFSET, RV_PLIC_IE00_E31,
+      },
+      {
+          RV_PLIC_IE01_REG_OFFSET, RV_PLIC_IE01_E63,
+      },
+      {
+          RV_PLIC_IE02_REG_OFFSET, RV_PLIC_IE02_E80,
+      },
+  };
+};
+
+TEST_F(IrqEnableSetTest, NullArgs) {
+  dif_plic_result_t result =
+      dif_plic_irq_enable_set(nullptr, kTopEarlgreyPlicIrqIdGpioGpio0,
+                              kTopEarlgreyPlicTargetIbex0, kDifPlicEnable);
```

Comment:
Given this is a DIF-specific mock test, I'm not sure we should be testing with `kTopEarlgreyPlic*` constants - is it maybe more obvious just to test with the target id or the interrupt id (in this case both 0?)? Then the test won't change if the earlgrey plic layout changes. You can also use the parameter definitions from `rv_plic_regs.h` rather than `kTopEarlgreyPlic*Last` for iteration.

I note you are already using some of the parameters from `rv_plic_regs.h` in `ExpectInitReset()`, above.

Reply:
Ill do that, but I don't really see anything criminal in using `kTopEarlgreyPlic*`.

What is your concern about using `kTopEarlgreyPlic*` ?

Reply:
> Then the test won't change if the earlgrey plic layout changes.

The DIF has been written not to rely on top_earlgrey, I think it makes sense that the mock tests (which don't require a full system), don't use the system defines either.

Reply:
I mean, it makes sense, however both `*regs.h` and system defines are generated from the same source. I think if one changes, the other one will as well? But anyway, I don't have anything against this change - will do.

Reply:
Done.